internal/broadcasters.go

```diff
@@ -48,27 +49,23 @@ func (b *Broadcaster[V]) AddListener() <-chan V {
 func (b *Broadcaster[V]) RemoveListener(ch <-chan V) {
 	b.lock.Lock()
 	defer b.lock.Unlock()
-	ss := b.subscribers
-	for i, s := range ss {
+	b.subscribers = slices.DeleteFunc(b.subscribers, func(pair channelPair[V]) bool {
 		// The following equality test is the reason why we have to store both the sendCh (chan X) and
 		// the receiveCh (<-chan X) for each subscriber; "s.sendCh == ch" would not be true because
 		// they're of two different types.
-		if s.receiveCh == ch {
-			copy(ss[i:], ss[i+1:])
-			ss[len(ss)-1] = channelPair[V]{}
-			b.subscribers = ss[:len(ss)-1]
-			close(s.sendCh)
-			break
+		if pair.receiveCh == ch {
+			close(pair.sendCh)
+			return true
 		}
-	}
+		return false
+	})
 }
 
 // HasListeners returns true if there are any current subscribers.
 func (b *Broadcaster[V]) HasListeners() bool {
 	b.lock.RLock()
-	hasListeners := len(b.subscribers) > 0
-	b.lock.RUnlock()
-	return hasListeners
+	defer b.lock.RUnlock()
+	return len(b.subscribers) > 0
```

Comment:
Explicitly locking/unlocking is probably a tiny bit faster, but I like the defer pattern here.

Reply:
That was my first thought as well, but the contribution guidelines told me not to 😉

Reply:
Ah, yea, I see that note. That's specifically referring to hot paths like the evaluation engine, but I see how it's rather ambiguous.